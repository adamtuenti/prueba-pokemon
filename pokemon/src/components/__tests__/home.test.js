import React from "react"
import { render, screen, fireEvent, waitFor } from "@testing-library/react"

import  Home  from '../home/home'


const testPokemon = {
    id: '026',
    name: 'Raichu',
    attack: 90,
    defense: 55,
    image: 'https://w7.pngwing.com/pngs/152/256/png-transparent-pikachu-raichu-pokemon-pichu-pikachu-food-cat-like-mammal-carnivoran.png'
};

jest.mock("../../services/api", () => ({
    getPokemons: () => Promise.resolve([testPokemon]),
    detelePokemon: () => Promise.resolve(),
}));


const defaultProps = {
    pokemones: [testPokemon],
    crearPokemon: false,
    labelFilter: '',
    accion: '',
    idPokemon: '',
    page: 2,
    slice: [testPokemon],
    rows: 5
}



const setUp = (props) => {
    return render(<Home {...defaultProps} {...props} />);
  };


describe("Home Page", () => {


    beforeEach(() => {
        jest.clearAllMocks();
      });

    afterEach(() => {
    window.history.pushState(null, document.title, '/')
    })

    it("Deberia renderizar homePage", async () => {
        const { baseElement } = setUp();

        await waitFor(() => {
            expect(screen.getByText(testPokemon.name)).toBeInTheDocument();
        });

        expect(baseElement).toMatchSnapshot();
        
    })

    it("Deberia ir a la pagina 1", async () => {
        setUp();
        await waitFor(() => {
            expect(screen.getByText(testPokemon.name)).toBeInTheDocument();
        });

        const paginationButton = screen.getByText('1');

        fireEvent.click(paginationButton);
    })

    it("Deberia renderdizar un pokemon", async () => {
        
        setUp();
        
        await waitFor(() => {
            expect(screen.getByText(testPokemon.name)).toBeInTheDocument();
        });

        const raichuAttack = screen.getByText(testPokemon.attack)
        const raichuDeffence = screen.getByText(testPokemon.defense)
        const raichuImage = screen.getByAltText(testPokemon.id);
        
        expect(raichuAttack).toBeInTheDocument();
        expect(raichuDeffence).toBeInTheDocument();
        expect(raichuImage.src).toContain(testPokemon.image);
        
    })

    it("Deberia editar un pokemon", async () => {
        
        setUp();
        
        await waitFor(() => {
            expect(screen.getByText(testPokemon.name)).toBeInTheDocument();
        });

        const editarPokemonButton = screen.getByTestId('editar-pokemon');
        fireEvent.click(editarPokemonButton);
        
    })

    it("Deberia eliminar un pokemon", async () => {
        
        setUp();
        
        await waitFor(() => {
            expect(screen.getByText(testPokemon.name)).toBeInTheDocument();
        });

        const eliminarPokemon = screen.getByTestId('eliminar-pokemon');
        fireEvent.click(eliminarPokemon);
        
        const tituloAdvertencia = screen.getByText('Â¿Deseas eliminar a '+ testPokemon.name +'?')
        expect(tituloAdvertencia).toBeInTheDocument();

        const confirmarEliminar = screen.getByText('Si');
        fireEvent.click(confirmarEliminar);
        
        await waitFor(() => {
            expect(screen.queryByText(testPokemon.name)).not.toBeInTheDocument();
        });
        
    })

    it("Deberia buscar un pokemon",  async () => {
        setUp();

        await waitFor(() => {
            expect(screen.getByText(testPokemon.name)).toBeInTheDocument();
        });
        
        const barraBusqueda = screen.getByPlaceholderText('Buscar');
        fireEvent.change(barraBusqueda, { target: { value: testPokemon.name } });
        expect(barraBusqueda.value).toStrictEqual(testPokemon.name);
        
    })

    it("Deberia abrir el formulario para un nuevo pokemon", () => {

        setUp();   
        const botonNuevo = screen.getByText('Nuevo');

        fireEvent.click(botonNuevo);
        
        const nuevoPokemonTitulo = screen.getByTestId('nuevo-pokemon');
        expect(nuevoPokemonTitulo).toBeInTheDocument(); 

    })
  })