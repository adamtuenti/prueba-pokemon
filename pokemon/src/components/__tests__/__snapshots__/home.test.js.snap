// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Deberia renderizar homePage 1`] = `
<body>
  <div>
    <div
      class="container"
      id="containerGeneral"
    >
      <p
        style="font-size: 25px; margin-bottom: 25px;"
      >
        Listado de Pokemones
      </p>
      <div
        class="row"
        style="margin-bottom: 24.5px;"
      >
        <div
          class="justify-content-end align-items-end col-lg-6 col-sm-6 col-12"
        >
          <div
            style="border: 0.5px solid; width: 62.5%; padding: 3.5px;"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="margin-left: 7.5px;"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
            <input
              placeholder="Buscar"
              style="padding-left: 7.5px; border: transparent; outline: none; width: 85%;"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="d-flex flex-col justify-content-end align-items-end col-lg-6 col-sm-6 col-12"
        >
          <button
            class="boton"
            data-testid="nuevo-pokemon"
          >
            <svg
              color="white"
              fill="currentColor"
              height="25"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 24 24"
              width="25"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
             Nuevo
          </button>
        </div>
      </div>
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                Nombre
              </th>
              <th>
                Imagen
              </th>
              <th>
                Ataque
              </th>
              <th>
                Defensa
              </th>
              <th>
                Acciones
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                Raichu
              </td>
              <td>
                <img
                  alt="026"
                  class="img-fluid"
                  src="https://w7.pngwing.com/pngs/152/256/png-transparent-pikachu-raichu-pokemon-pichu-pikachu-food-cat-like-mammal-carnivoran.png"
                  style="width: 45px;"
                />
              </td>
              <td>
                90
              </td>
              <td>
                55
              </td>
              <td>
                <div>
                  <svg
                    color="#847cdc"
                    data-testid="eliminar-pokemon"
                    fill="currentColor"
                    height="23.5"
                    stroke="currentColor"
                    stroke-width="0"
                    style="color: rgb(132, 124, 220);"
                    viewBox="0 0 24 24"
                    width="23.5"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0V0z"
                      fill="none"
                    />
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                  <svg
                    color="#847cdc"
                    data-testid="editar-pokemon"
                    fill="currentColor"
                    height="23.5"
                    stroke="currentColor"
                    stroke-width="0"
                    style="color: rgb(132, 124, 220); margin-left: 7.5px;"
                    viewBox="0 0 24 24"
                    width="23.5"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M22 24H2v-4h20v4zM13.06 5.19l3.75 3.75L7.75 18H4v-3.75l9.06-9.06zm4.82 2.68l-3.75-3.75 1.83-1.83a.996.996 0 011.41 0l2.34 2.34c.39.39.39 1.02 0 1.41l-1.83 1.83z"
                    />
                  </svg>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        <button
          class="activeButton button"
        >
          1
        </button>
      </div>
    </div>
  </div>
</body>
`;
